core/dbt/task/run.py

```diff
@@ -284,18 +374,95 @@ def execute(self, model, manifest):
             )
 
         hook_ctx = self.adapter.pre_model_hook(context_config)
+        batch_results = None
         try:
-            result = MacroGenerator(
-                materialization_macro, context, stack=context["context_macro_stack"]
-            )()
+            if (
+                os.environ.get("DBT_EXPERIMENTAL_MICROBATCH")
+                and model.config.materialized == "incremental"
+                and model.config.incremental_strategy == "microbatch"
+            ):
+                batch_results = self._execute_microbatch_materialization(
+                    model, manifest, context, materialization_macro
+                )
+            else:
+                result = MacroGenerator(
+                    materialization_macro, context, stack=context["context_macro_stack"]
+                )()
+                for relation in self._materialization_relations(result, model):
+                    self.adapter.cache_added(relation.incorporate(dbt_created=True))
         finally:
             self.adapter.post_model_hook(context_config, hook_ctx)
 
-        for relation in self._materialization_relations(result, model):
-            self.adapter.cache_added(relation.incorporate(dbt_created=True))
+        if batch_results:
+            return self._build_run_microbatch_model_result(model, batch_results)
 
         return self._build_run_model_result(model, context)
 
+    def _execute_microbatch_materialization(self, model, manifest, context, materialization_macro):
+        batch_results = []
+        microbatch_builder = MicrobatchBuilder(
+            model=model,
+            is_incremental=self._is_incremental(model),
+            event_time_start=getattr(self.config.args, "EVENT_TIME_START", None),
+            event_time_end=getattr(self.config.args, "EVENT_TIME_END", None),
+        )
+        end = microbatch_builder.build_end_time()
+        start = microbatch_builder.build_start_time(end)
```

Comment:
When I was reading through `MicrobatchBuilder` I found it odd that `build_end_time` and `build_start_time` weren't private functions that the `__init__` would then call to default the `event_time_start`/`event_time_end`. With this bit of code here, I find myself still thinking so. Basically, the start/end are always going to be specific to a given `MicrobatchBuilder` instance. Perhaps as a fast follow to this PR we should investigate if this can be reduced to that. The code here would then become
```
...
microbatch_builder = MicrobatchBuilder(
  model=model,
  is_incremental=self._is_incremental(model),
  event_time_start=getattr(self.config.args, "EVENT_TIME_START", None),
  event_time_end=getattr(self.config.args, "EVENT_TIME_END", None),
)
batches = microbatch_builder.build_batches()
...
```

The alternative would be to take the `MicrobatchBuilder` class to be less specific to a model, which also seems like a valid approach. Right now though we seem to be somewhere in the middle with a class that is specific to a model, but methods that we're expected to call which should never change their return given the associated model.